File: src/siminov/connect/sync/SyncWorker.java
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package siminov.connect.sync;

import java.util.Collection;
import java.util.Iterator;
import java.util.concurrent.ConcurrentLinkedQueue;

import siminov.connect.Constants;
import siminov.connect.IWorker;
import siminov.connect.events.ISyncEvents;
import siminov.connect.model.ServiceDescriptor;
import siminov.connect.model.ServiceDescriptor.API;
import siminov.connect.model.SyncDescriptor;
import siminov.connect.resource.Resources;
import siminov.connect.service.NameValuePair;
import siminov.connect.service.design.IService;
import siminov.connect.sync.design.ISyncRequest;
import siminov.orm.utils.ClassUtils;

public class SyncWorker implements IWorker {

	private static SyncWorker syncWorker = null;
	private SyncWorkerThread syncWorkerThread = null;
	
	private Collection<ISyncRequest> syncRequests = new ConcurrentLinkedQueue<ISyncRequest>();
	
	private Resources resources = Resources.getInstance();
	
	private SyncWorker() {
		
	}

	public static SyncWorker getInstance() {
		
		if(syncWorker == null) {
			syncWorker = new SyncWorker();
		}
		
		return syncWorker;
	}
	
	
	public void startWorker() {
		
		if(syncWorkerThread == null) {
			syncWorkerThread = new SyncWorkerThread();
		}
		
		syncWorkerThread.start();
	}
	
	public void stopWorker() {
		
		if(syncWorkerThread == null) {
			return;
		}
		
		
		if(!syncWorkerThread.isAlive()) {
			syncWorkerThread.start();
		}
	}
	
	public boolean isWorkerRunning() {
		
		if(syncWorkerThread == null) {
			return false;
		}
		
		return syncWorkerThread.isAlive();
	}
	
	
	private class SyncWorkerThread extends Thread {
		
		public void run() {
			
			Iterator<ISyncRequest> requests = syncRequests.iterator();
			while(requests.hasNext()) {
				
				ISyncRequest syncRequest = requests.next();

				/*
				 * Fire Sync Started Event
				 */
				ISyncEvents syncEventHandler = resources.getSyncEventHandler();
				if(syncEventHandler != null) {
					syncEventHandler.onStart(syncRequest);
				}
				
				
				/*
				 * Process Request
				 */
				SyncDescriptor refreshDescriptor = resources.getSyncDescriptor(syncRequest.getName());
				
				Iterator<String> services = refreshDescriptor.getServices();
				while(services.hasNext()) {
					
					String service = services.next();
					
					String serviceName = service.substring(0, service.indexOf(Constants.SYNC_DESCRIPTOR_SERVICE_SEPARATOR));
					String apiName = service.substring(service.indexOf(Constants.SYNC_DESCRIPTOR_SERVICE_SEPARATOR) + 1, service.length());

					
					ServiceDescriptor serviceDescriptor = resources.requiredServiceDescriptorBasedOnName(serviceName);
					API api = serviceDescriptor.getApi(apiName);
					
					String apiHandler = api.getHandler();
					
					IService serviceHandler = (IService) ClassUtils.createClassInstance(apiHandler);
					serviceHandler.setServiceDescriptor(serviceDescriptor);

					Iterator<NameValuePair> resources = syncRequest.getResources();
					while(resources.hasNext()) {
						serviceHandler.addResource(resources.next());
					}
					
					
					serviceHandler.invoke();
				}
				
				
				/*
				 * Fire Sync Started Event
				 */
				if(syncEventHandler != null) {
					syncEventHandler.onFinish(syncRequest);
				}
				
				
				syncRequests.remove(syncRequest);
			}
			
			
			syncWorkerThread = null;
		}
	}
	

	public void addRequest(final ISyncRequest syncRequest) {
		
		if(containRequest(syncRequest)) {
			return;
		}
		
		
		this.syncRequests.add(syncRequest);

		/*
		 * Fire Sync Queued Event
		 */
		ISyncEvents syncEventHandler = resources.getSyncEventHandler();
		if(syncEventHandler != null) {
			syncEventHandler.onQueue(syncRequest);
		}
		
		
		if(!isWorkerRunning()) {
			startWorker();
		}
	}
	
	public boolean containRequest(final ISyncRequest refreshRequest) {
		return this.syncRequests.contains(refreshRequest);
	}
	
	public void removeRequest(final ISyncRequest refreshRequest) {
		this.syncRequests.remove(refreshRequest);
	}
}

File: Windows/universal8.1/Siminov/Connect/Sync/SyncWorker.cs
///
/// [SIMINOV FRAMEWORK - CONNECT]
/// Copyright [2014-2016] [Siminov Software Solution LLP|support@siminov.com]
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.



using Siminov.Connect;
using Siminov.Connect.Events;
using Siminov.Connect.Model;
using Siminov.Connect.Resource;
using Siminov.Connect.Service.Design;
using Siminov.Connect.Sync.Design;
using Siminov.Core.Utils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace connect.Siminov.Connect.Sync
{

    /// <summary>
    /// It provides implementation for Sync IWorker
    /// It processes all sync requests
    /// </summary>
    public class SyncWorker : IWorker
    {

	    private static SyncWorker syncWorker = null;
	    private static Task syncWorkerThread = null;
	
	    private static ICollection<ISyncRequest> syncRequests = new LinkedList<ISyncRequest>();
	
	    private static ResourceManager resourceManager = ResourceManager.GetInstance();
	
        /// <summary>
        /// SyncWorker Private Constructor
        /// </summary>
	    private SyncWorker() 
        {
		
	    }

        /// <summary>
        /// It provides singleton instance of SyncWorker
        /// </summary>
        /// <returns>Singleton instance of SyncWorker</returns>
	    public static SyncWorker GetInstance() 
        {
		
		    if(syncWorker == null) 
            {
			    syncWorker = new SyncWorker();
		    }
		
		    return syncWorker;
	    }
	
	
	    public void StartWorker() 
        {
		
		    if(syncWorkerThread == null) 
            {
			    syncWorkerThread = new Task(new Action(HandleRequests));
		    }
		
		    syncWorkerThread.Start();
	    }
	
	    public void StopWorker() 
        {
		
		    if(syncWorkerThread == null) 
            {
			    return;
		    }
		
		
		    //if(!syncWorkerThread.IsAlive()) 
            {
			    syncWorkerThread.Start();
		    }
	    }
	
	    public bool IsWorkerRunning() 
        {
		
		    if(syncWorkerThread == null) 
            {
			    return false;
		    }
		
		    return syncWorkerThread.IsCompleted;
	    }


        public void RemoveRequest(IRequest request)
        {

        }


	    public void HandleRequests() 
        {
			
			IEnumerator<ISyncRequest> requests = syncRequests.GetEnumerator();
            ICollection<ISyncRequest> removeRequests = new List<ISyncRequest>();

			while(requests.MoveNext()) 
            {
				
				ISyncRequest syncRequest = requests.Current;

				/*
				    * Fire Sync Started Event
				    */
				ISyncEvents syncEventHandler = resourceManager.GetSyncEventHandler();
				if(syncEventHandler != null) 
                {
					syncEventHandler.OnStart(syncRequest);
				}
				
				
				/*
				    * Process Request
				    */
				SyncDescriptor refreshDescriptor = resourceManager.GetSyncDescriptor(syncRequest.GetName());
				
				IEnumerator<String> services = refreshDescriptor.GetServiceDescriptorNames();
				while(services.MoveNext()) 
                {
					
					String service = services.Current;
					
					String serviceName = service.Substring(0, service.IndexOf(Constants.SYNC_DESCRIPTOR_SERVICE_SEPARATOR));
                    String requestName = service.Substring(service.IndexOf(Constants.SYNC_DESCRIPTOR_SERVICE_SEPARATOR) + 1, service.Length - (service.IndexOf(Constants.SYNC_DESCRIPTOR_SERVICE_SEPARATOR) + 1));

					
					ServiceDescriptor serviceDescriptor = resourceManager.RequiredServiceDescriptorBasedOnName(serviceName);
					ServiceDescriptor.Request api = serviceDescriptor.GetRequest(requestName);
					
					String apiHandler = api.GetHandler();
					
					IService serviceHandler = (IService) ClassUtils.CreateClassInstance(apiHandler);
					serviceHandler.SetServiceDescriptor(serviceDescriptor);

					IEnumerator<String> resources = syncRequest.GetResources();
					while(resources.MoveNext()) 
                    {
						String resourceName = resources.Current;
						Object resourceValue = syncRequest.GetResource(resourceName);
						
						serviceHandler.AddResource(resourceName, resourceValue);
					}
					
					
					serviceHandler.Invoke();
				}
				
				
				/*
				    * Fire Sync Started Event
				    */
				if(syncEventHandler != null) {
					syncEventHandler.OnFinish(syncRequest);
				}
				
				
				removeRequests.Add(syncRequest);
			}
			
			foreach(ISyncRequest removeRequest in removeRequests) 
            {
                syncRequests.Remove(removeRequest);
            }

			syncWorkerThread = null;
		}
	

	    public void AddRequest(IRequest request) 
        {
		
		    ISyncRequest syncRequest = (ISyncRequest) request;
		    if(ContainsRequest(syncRequest)) 
            {
			    return;
		    }
		
		
		    syncRequests.Add(syncRequest);

		    /*
		     * Fire Sync Queued Event
		     */
		    ISyncEvents syncEventHandler = resourceManager.GetSyncEventHandler();
		    if(syncEventHandler != null) 
            {
			    syncEventHandler.OnQueue(syncRequest);
		    }
		
		
		    if(!IsWorkerRunning()) 
            {
			    StartWorker();
		    }
	    }
	
	    public bool ContainsRequest(IRequest refreshRequest) 
        {
		    return syncRequests.Contains(refreshRequest);
	    }
	
	    public void removeRequest(IRequest refreshRequest) 
        {
		    syncRequests.Remove((ISyncRequest) refreshRequest);
	    }
    }
}
